packages/realm-web/src/tests/App.test.ts

```diff
@@ -489,6 +478,66 @@ describe("App", () => {
                     Authorization: "Bearer very-refreshing",
                 },
             },
+            {
+                method: "POST",
+                url:
+                    "http://localhost:1337/api/client/v2.0/app/my-mocked-app/functions/call",
+                body: { name: "foo", arguments: [{ bar: "baz" }] },
+                headers: {
+                    ...SENDING_JSON_HEADERS,
+                    Authorization: "Bearer second-access-token",
+                },
+            },
+        ]);
+    });
+
+    it("attempts to refresh access token, retries request exacly once, upon an 'invalid session' (401) response", async () => {
+        const invalidSessionError = new MongoDBRealmError(
+            "POST",
+            "http://localhost:1337/some-path",
+            401,
+            "",
+            "invalid session",
+        );
+        const app = new MockApp("my-mocked-app", [
+            LOCATION_RESPONSE,
+            {
+                user_id: "bobs-id",
+                access_token: "first-access-token",
+                refresh_token: "very-refreshing",
+            },
+            invalidSessionError,
+            invalidSessionError,
+            invalidSessionError,
+        ]);
+        // Login with an anonymous user
+        const credentials = Credentials.anonymous();
+        const user = await app.logIn(credentials, false);
+        // Send a request (which will fail)
+        try {
+            await app.functions.foo({ bar: "baz" });
+            throw new Error("Expected the request to fail");
+        } catch (err) {
+            expect(err).instanceOf(MongoDBRealmError);
+            if (err instanceof MongoDBRealmError) {
+                expect(err.message).equals(
+                    // The last failure is from failing to delete the session at logout
+                    "Request failed (DELETE http://localhost:1337/api/client/v2.0/auth/session): invalid session (status 401)",
+                );
+            }
+        }
```

Comment:
Maybe we should consider adding [chai-as-promised](https://www.npmjs.com/package/chai-as-promised)?

Reply:
Yeah - I considered adding it. I don't like they way it gets registered into chai, but I might be able to overcome that if I keep repeating this pattern.